File: ANTLR4/Calculadora/EvalVisitor.py
from LabeledExprVisitor import LabeledExprVisitor
from LabeledExprParser import LabeledExprParser
import math

class EvalVisitor(LabeledExprVisitor):
    def __init__(self):
        self.memory = {}

    def visitAssign(self, ctx):
        id_name = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id_name] = value
        return value

    def visitPrintExpr(self, ctx):
        value = self.visit(ctx.expr())
        print(value)
        return 0.0

    def visitInt(self, ctx):
        return float(ctx.INT().getText())

    def visitId(self, ctx):
        id_name = ctx.ID().getText()
        return self.memory.get(id_name, 0.0)

    def visitMulDiv(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.MUL:
            return left * right
        else:  # DIV
            if right == 0:
                print("División por cero no se puede ")
                return 0.0
            return left / right


    def visitAddSub(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        return left + right if ctx.op.type == LabeledExprParser.ADD else left - right

    def visitParens(self, ctx):
        return self.visit(ctx.expr())

    def _is_radians(self, ctx):
        return hasattr(ctx, "prefix") and ctx.prefix is not None and ctx.prefix.text.lower() == "rad"

    def visitSin(self, ctx):
        value = self.visit(ctx.expr())
        is_radians = ctx.prefix is not None and ctx.prefix.text.lower() == "rad"
        return math.sin(value if is_radians else math.radians(value))


    def visitCos(self, ctx):
        value = self.visit(ctx.expr())
        is_radians = ctx.prefix is not None and ctx.prefix.text.lower() == "rad"
        return math.cos(value if self._is_radians(ctx) else math.radians(value))

    def visitTan(self, ctx):
        value = self.visit(ctx.expr())
        is_radians = ctx.prefix is not None and ctx.prefix.text.lower() == "rad"
        return math.tan(value if self._is_radians(ctx) else math.radians(value))


    def visitSqrt(self, ctx):
        return math.sqrt(self.visit(ctx.expr()))

    def visitLn(self, ctx):
        return math.log(self.visit(ctx.expr()))

    def visitLog(self, ctx):
        return math.log10(self.visit(ctx.expr()))

    def visitFactorial(self, ctx):
        raw = self.visit(ctx.expr())
        n = int(raw)

        if raw != n or n < 0:
            print(f"Negativo o decimal no se puede {raw}")
            return raw

        result = 1.0
        for i in range(2, n + 1):
            result *= i
        return result

File: ANTLR4/Calculadora/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Double> {
    Map<String, Double> memory = new HashMap<>();
	
	/** ID '=' expr NEWLINE */
	@Override
	public Double visitAssign(LabeledExprParser.AssignContext ctx) {
		String id = ctx.ID().getText(); // id is left-hand side of '='
		double value = visit(ctx.expr()); // compute value of expression on right
		memory.put(id, value); // store it in our memory
		return value;
	}
	/** expr NEWLINE */
	@Override
	public Double visitPrintExpr(LabeledExprParser.PrintExprContext ctx) {
		Double value = visit(ctx.expr()); // evaluate the expr child
		System.out.println(value); // print the result
		return 0.0; // return dummy value
	}
	
	/** INT */
	@Override
	public Double visitInt(LabeledExprParser.IntContext ctx) {
		return Double.valueOf(ctx.INT().getText());
	}
	
	/** ID */
	@Override
	public Double visitId(LabeledExprParser.IdContext ctx) {
		String id = ctx.ID().getText();
		if ( memory.containsKey(id) ) return memory.get(id);
		return 0.0;
	}
	
	/** expr op=('*'|'/') expr */
	@Override
  public Double visitMulDiv(LabeledExprParser.MulDivContext ctx) {
      double left = visit(ctx.expr(0));
      double right = visit(ctx.expr(1));
      return ctx.op.getType() == LabeledExprParser.MUL ? left * right : left / right;
  }
	
	/** expr op=('+'|'-') expr */
	@Override
	public Double visitAddSub(LabeledExprParser.AddSubContext ctx) {
    double left = visit(ctx.expr(0));
    double right = visit(ctx.expr(1));
    return ctx.op.getType() == LabeledExprParser.ADD ? left + right : left - right;
  }
	
	/** '(' expr ')' */
	@Override
	public Double visitParens(LabeledExprParser.ParensContext ctx) {
		return visit(ctx.expr()); // return child expr's value
	}
	
	@Override
  public Double visitSin(LabeledExprParser.SinContext ctx) {
      double value = visit(ctx.expr());
      boolean isRad = ctx.prefix != null && ctx.prefix.getText().equalsIgnoreCase("rad");
      return isRad ? Math.sin(value) : Math.sin(Math.toRadians(value));
  }

  @Override
  public Double visitCos(LabeledExprParser.CosContext ctx) {
      double value = visit(ctx.expr());
      boolean isRad = ctx.prefix != null && ctx.prefix.getText().equalsIgnoreCase("rad");
      return isRad ? Math.cos(value) : Math.cos(Math.toRadians(value));
  }

  @Override
  public Double visitTan(LabeledExprParser.TanContext ctx) {
      double value = visit(ctx.expr());
      boolean isRad = ctx.prefix != null && ctx.prefix.getText().equalsIgnoreCase("rad");
      return isRad ? Math.tan(value) : Math.tan(Math.toRadians(value));
  }


  @Override
  public Double visitSqrt(LabeledExprParser.SqrtContext ctx) {
      return Math.sqrt(visit(ctx.expr()));
  }

  @Override
  public Double visitLn(LabeledExprParser.LnContext ctx) {
      return Math.log(visit(ctx.expr()));
  }

  @Override
  public Double visitLog(LabeledExprParser.LogContext ctx) {
      return Math.log10(visit(ctx.expr()));
  }

  @Override
  public Double visitFactorial(LabeledExprParser.FactorialContext ctx) {
      double raw = visit(ctx.expr());
      int n = (int) raw;

      if (raw != n || n < 0) {
          System.out.println("Negativo o decimal no se puede, tonto ");
          return raw;
      }

      double result = 1.0;
      for (int i = 2; i <= n; i++) result *= i;
      return result;
  }


}

